tor/controller.go

```diff
@@ -0,0 +1,471 @@
+// be the ports where the onion service can be reached at, while the mapped
+// target ports will be the ports where the onion service is running locally.
+// For example, if you wish to run an onion service that should be accessed on
+// port 9735, but the service is listening on 9000, 9001, and 9002 internally,
+// then a mapping of {9735: {9001, 9002, 9003}} can be provided.
+func (c *Controller) AddOnionV2(privateKeyFile string,
+	virtToTargPorts map[int][]int) ([]*OnionAddr, error) {
+
+	// We'll start off by reading the private key from its path.
+	privateKey, err := ioutil.ReadFile(privateKeyFile)
+	if err != nil && !strings.Contains(err.Error(), "no such file") {
+		return nil, fmt.Errorf("unable to read private key from file: "+
+			"%v", err)
+	}
+
+	// If the file does not exist, then we should request the server to
+	// create a new onion service and return its private key. Otherwise,
+	// we'll request the server to recreate the onion server from our
+	// private key.
+	var keyParam string
+	if privateKey != nil {
+		keyParam = string(privateKey)
+	} else {
+		keyParam = "NEW:RSA1024"
+	}
+
+	// Now, we'll determine the different virtual ports on which this onion
+	// service will be accessed by.
+	var portParam string
+	for virtPort, targPorts := range virtToTargPorts {
+		// If the virtual port doesn't map to any target ports, we'll
+		// use the virtual port as the target port.
+		if len(targPorts) == 0 {
+			portParam += fmt.Sprintf("Port=%d,%d ", virtPort,
+				virtPort)
+			continue
+		}
+
+		// Otherwise, we'll create a mapping from the virtual port to
+		// each target port.
+		for _, targPort := range targPorts {
+			portParam += fmt.Sprintf("Port=%d,%d ", virtPort,
+				targPort)
+		}
+	}
+
+	cmd := fmt.Sprintf("ADD_ONION %s %s", keyParam, portParam)
+	_, reply, err := c.sendCommand(cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	// If successful, the reply from the server should be of the following
+	// format, depending on whether a private key has been requested:
+	//
+	//	C: ADD_ONION NEW:RSA1024 Port=80,8080
+	//	S: 250-ServiceID=testonion1234567
+	//	S: 250 OK
+	//
+	//	C: ADD_ONION RSA1024:[Blob Redacted] Port=80,8080
+	//	S: 250-ServiceID=testonion1234567
+	//	S: 250-PrivateKey=RSA1024:[Blob Redacted]
+	//	S: 250 OK
+	//
+	// We're interested in retrieving the service ID, which is the public
+	// name of the service, and the private key if requested.
+	replyParams := parseTorReply(reply)
+	serviceID, exists := replyParams["ServiceID"]
+	if !exists {
+		return nil, errors.New("service id not found in reply")
+	}
+
+	// If a new onion service was created, we'll write its private key to
+	// disk under strict permissions in the event that it needs to be
+	// recreated later on.
+	if privateKey == nil {
+		privateKey, exists := replyParams["PrivateKey"]
+		if !exists {
+			return nil, errors.New("private key not found in reply")
+		}
+
+		err := ioutil.WriteFile(privateKeyFile, []byte(privateKey), 0600)
```

Comment:
Given the comment above, perhaps we should encode this in the proper format, so we don't need to do any special decoding on start up.

Reply:
See comment above.